src/rsz/src/Resizer.cc

```diff
@@ -762,35 +904,82 @@ bool Resizer::isTristateDriver(const Pin* pin)
 void Resizer::bufferOutput(const Pin* top_pin, LibertyCell* buffer_cell)
 {
   NetworkEdit* network = networkEdit();
-  Term* term = network_->term(top_pin);
-  Net* output_net = network_->net(term);
+
+  odb::dbITerm* top_pin_op_iterm;
+  odb::dbBTerm* top_pin_op_bterm;
+  odb::dbModITerm* top_pin_op_moditerm;
+  odb::dbModBTerm* top_pin_op_modbterm;
+
+  db_network_->staToDb(top_pin,
+                       top_pin_op_iterm,
+                       top_pin_op_bterm,
+                       top_pin_op_moditerm,
+                       top_pin_op_modbterm);
+
+  //  Term* term = network_->term(top_pin);
+  odb::dbNet* flat_op_net = top_pin_op_bterm->getNet();
+  odb::dbModNet* hier_op_net = top_pin_op_bterm->getModNet();
+
+  sta_->disconnectPin(const_cast<Pin*>(top_pin));
+
   LibertyPort *input, *output;
   buffer_cell->bufferPorts(input, output);
+
   string buffer_name = makeUniqueInstName("output");
+  Net* buffer_out = makeUniqueNet();
   Instance* parent = network->topInstance();
-  Net* buffer_in = makeUniqueNet();
+
   Point pin_loc = db_network_->location(top_pin);
+  // buffer made in top level.
   Instance* buffer
       = makeBuffer(buffer_cell, buffer_name.c_str(), parent, pin_loc);
   inserted_buffer_count_++;
 
-  NetConnectedPinIterator* pin_iter
-      = network_->connectedPinIterator(output_net);
-  while (pin_iter->hasNext()) {
-    const Pin* pin = pin_iter->next();
-    if (pin != top_pin) {
-      // Leave output port pin connected to output_net.
-      sta_->disconnectPin(const_cast<Pin*>(pin));
-      Port* pin_port = network->port(pin);
-      sta_->connectPin(network->instance(pin), pin_port, buffer_in);
-    }
-  }
-  delete pin_iter;
-  sta_->connectPin(buffer, input, buffer_in);
-  sta_->connectPin(buffer, output, output_net);
-
-  parasiticsInvalid(buffer_in);
-  parasiticsInvalid(output_net);
+  // connect original input (hierarchical or flat) to buffer input
+  // handle hierarchy
+  Pin* buffer_ip_pin;
+  Pin* buffer_op_pin;
+  getBufferPins(buffer, buffer_ip_pin, buffer_op_pin);
+  odb::dbITerm* buffer_op_pin_iterm;
+  odb::dbBTerm* buffer_op_pin_bterm;
+  odb::dbModITerm* buffer_op_pin_moditerm;
+  odb::dbModBTerm* buffer_op_pin_modbterm;
+  db_network_->staToDb(buffer_op_pin,
+                       buffer_op_pin_iterm,
+                       buffer_op_pin_bterm,
+                       buffer_op_pin_moditerm,
+                       buffer_op_pin_modbterm);
+
+  odb::dbITerm* buffer_ip_pin_iterm;
+  odb::dbBTerm* buffer_ip_pin_bterm;
+  odb::dbModITerm* buffer_ip_pin_moditerm;
+  odb::dbModBTerm* buffer_ip_pin_modbterm;
+  db_network_->staToDb(buffer_ip_pin,
+                       buffer_ip_pin_iterm,
+                       buffer_ip_pin_bterm,
+                       buffer_ip_pin_moditerm,
+                       buffer_ip_pin_modbterm);
```

Comment:
This idiom occurs so often that it seems having a struct for the purpose would be helpful, ie
```
DbPin db_buffer_ip_pin;
db_network_->staToDb(buffer_ip_pin, db_buffer_ip_pin);
```

Reply:
I agree it is a little verbose, and there are many cases (eg when handling inserted liberty cells or even module instances) when we know the expected type of a pin. In those case there are some api calls to get the expected pin type.

For example, when we know that a pin is just an iterm we can use the getFlatPin.. Code updated accordingly. In this case we know that the type of the liberty buffer pin is always an iterm. Cleaned up.

 getBufferPins(buffer, buffer_ip_pin, buffer_op_pin);
  buffer_ip_iterm = db_network_ -> flatPin(buffer_ip_pin);
  buffer_op_iterm = db_network_ -> flatPin(buffer_op_pin);